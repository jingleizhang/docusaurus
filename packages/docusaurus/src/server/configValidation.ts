/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import {DocusaurusConfig} from '@docusaurus/types';
import Joi from '@hapi/joi';
import {CONFIG_FILE_NAME} from '../constants';

export const DEFAULT_CONFIG: Pick<
  DocusaurusConfig,
  | 'onBrokenLinks'
  | 'plugins'
  | 'themes'
  | 'presets'
  | 'customFields'
  | 'themeConfig'
> = {
  onBrokenLinks: 'throw',
  plugins: [],
  themes: [],
  presets: [],
  customFields: {},
  themeConfig: {},
};

const PluginSchema = Joi.alternatives().try(
  Joi.string(),
  Joi.array().items(Joi.string().required(), Joi.object().required()).length(2),
);

const ThemeSchema = Joi.alternatives().try(
  Joi.string(),
  Joi.array().items(Joi.string().required(), Joi.object().required()).length(2),
);

const PresetSchema = Joi.alternatives().try(
  Joi.string(),
  Joi.array().items(Joi.string().required(), Joi.object().required()).length(2),
);

const ConfigSchema = Joi.object({
  baseUrl: Joi.string()
    .required()
    .regex(new RegExp('/$', 'm'))
    .message('{{#label}} must be a string with a trailing `/`'),
  favicon: Joi.string().required(),
  title: Joi.string().required(),
  url: Joi.string().uri().required(),
  onBrokenLinks: Joi.string()
    .equal('ignore', 'log', 'error', 'throw')
    .default(DEFAULT_CONFIG.onBrokenLinks),
  organizationName: Joi.string(),
  projectName: Joi.string(),
  customFields: Joi.object().unknown().default(DEFAULT_CONFIG.customFields),
  githubHost: Joi.string(),
  plugins: Joi.array().items(PluginSchema).default(DEFAULT_CONFIG.plugins),
  themes: Joi.array().items(ThemeSchema).default(DEFAULT_CONFIG.themes),
  presets: Joi.array().items(PresetSchema).default(DEFAULT_CONFIG.presets),
  themeConfig: Joi.object().unknown().default(DEFAULT_CONFIG.themeConfig),
  scripts: Joi.array().items(
    Joi.string(),
    Joi.object({
      src: Joi.string().required(),
      async: Joi.bool(),
      defer: Joi.bool(),
    }).oxor('async', 'defer'),
  ),
  stylesheets: Joi.array().items(
    Joi.string(),
    Joi.object({
      href: Joi.string().uri().required(),
      type: Joi.string().required(),
    }),
  ),
  tagline: Joi.string(),
});

export function validateConfig(
  config: Partial<DocusaurusConfig>,
): DocusaurusConfig {
  const {error, value} = ConfigSchema.validate(config, {
    abortEarly: false,
  });
  if (error) {
    const unknownFields = error.details.reduce((formattedError, err) => {
      if (err.type === 'object.unknown') {
        return `${formattedError}"${err.path}",`;
      }
      return formattedError;
    }, '');
    let formattedError = error.details.reduce(
      (accumulatedErr, err) =>
        err.type !== 'object.unknown'
          ? `${accumulatedErr}${err.message}\n`
          : accumulatedErr,
      '',
    );
    formattedError = unknownFields
      ? `${formattedError}These field(s) [${unknownFields}] are not recognized in ${CONFIG_FILE_NAME}.\nIf you still want these fields to be in your configuration, put them in the 'customFields' attribute.\nSee https://v2.docusaurus.io/docs/docusaurus.config.js/#customfields`
      : formattedError;
    throw new Error(formattedError);
  } else {
    return value;
  }
}
