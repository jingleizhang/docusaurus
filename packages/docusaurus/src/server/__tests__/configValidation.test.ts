/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import {DEFAULT_CONFIG, validateConfig} from '../configValidation';
import {DocusaurusConfig} from '@docusaurus/types';

const baseConfig = {
  baseUrl: '/',
  favicon: 'some.ico',
  title: 'my site',
  url: 'https://mysite.com',
};

const testConfig = (config) => validateConfig({...baseConfig, ...config});

describe('validateConfig', () => {
  test('normalize config', () => {
    const value = testConfig({});
    expect(value).toEqual({
      ...DEFAULT_CONFIG,
      ...baseConfig,
    });
  });

  test('throw error for unknown field', () => {
    expect(() => {
      testConfig({
        invalid: true,
      });
    }).toThrowErrorMatchingSnapshot();
  });

  test('throw error for baseUrl without trailing `/`', () => {
    expect(() => {
      testConfig({
        baseUrl: 'noslash',
      });
    }).toThrowErrorMatchingSnapshot();
  });

  test('throw error if plugins is not array', () => {
    expect(() => {
      testConfig({
        plugins: {},
      });
    }).toThrowErrorMatchingSnapshot();
  });

  test('throw error if themes is not array', () => {
    expect(() => {
      testConfig({
        themes: {},
      });
    }).toThrowErrorMatchingSnapshot();
  });

  test('throw error if presets is not array', () => {
    expect(() => {
      testConfig({
        presets: {},
      });
    }).toThrowErrorMatchingSnapshot();
  });

  test("throw error if scripts doesn't have src", () => {
    expect(() => {
      testConfig({
        scripts: ['https://some.com', {}],
      });
    }).toThrowErrorMatchingSnapshot();
  });

  test("throw error if css doesn't have href", () => {
    expect(() => {
      testConfig({
        stylesheets: ['https://somescript.com', {type: 'text/css'}],
      });
    }).toThrowErrorMatchingSnapshot();
  });

  test('custom field in config', () => {
    const value = testConfig({
      customFields: {
        author: 'anshul',
      },
    });
    expect(value).toEqual({
      ...DEFAULT_CONFIG,
      ...baseConfig,
      customFields: {
        author: 'anshul',
      },
    });
  });

  test('throw error for required fields', () => {
    expect(
      () =>
        validateConfig(({
          invalid: true,
          preset: {},
          stylesheets: {},
          themes: {},
          scripts: {},
        } as unknown) as DocusaurusConfig), // to fields not in the type
    ).toThrowErrorMatchingSnapshot();
  });
});
