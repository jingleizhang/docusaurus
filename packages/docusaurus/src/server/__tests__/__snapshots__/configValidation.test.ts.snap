// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`validateConfig throw error for baseUrl without trailing \`/\` 1`] = `
"\\"baseUrl\\" must be a string with a trailing \`/\`
"
`;

exports[`validateConfig throw error for required fields 1`] = `
"\\"baseUrl\\" is required
\\"favicon\\" is required
\\"title\\" is required
\\"url\\" is required
\\"themes\\" must be an array
\\"scripts\\" must be an array
\\"stylesheets\\" must be an array
These field(s) [\\"invalid\\",\\"preset\\",] are not recognized in docusaurus.config.js.
If you still want these fields to be in your configuration, put them in the 'customFields' attribute.
See https://v2.docusaurus.io/docs/docusaurus.config.js/#customfields"
`;

exports[`validateConfig throw error for unknown field 1`] = `
"These field(s) [\\"invalid\\",] are not recognized in docusaurus.config.js.
If you still want these fields to be in your configuration, put them in the 'customFields' attribute.
See https://v2.docusaurus.io/docs/docusaurus.config.js/#customfields"
`;

exports[`validateConfig throw error if css doesn't have href 1`] = `
"\\"stylesheets[1]\\" does not match any of the allowed types
"
`;

exports[`validateConfig throw error if plugins is not array 1`] = `
"\\"plugins\\" must be an array
"
`;

exports[`validateConfig throw error if presets is not array 1`] = `
"\\"presets\\" must be an array
"
`;

exports[`validateConfig throw error if scripts doesn't have src 1`] = `
"\\"scripts[1]\\" does not match any of the allowed types
"
`;

exports[`validateConfig throw error if themes is not array 1`] = `
"\\"themes\\" must be an array
"
`;
